File: static/js/model/state-tracker.js
var DUB = DUB || {};
DUB.model = DUB.model || {};

(function (ns) {
	var EMPTY = -1;

	ns.StateTracker = function () {
		var MAX_STATES = 250;
		var stateStack = [];
		
		return {
			forward: function (state) {
				if (typeof state === "string") {
					stateStack.push(state);
					if (stateStack.length > 250) {
						stateStack.shift();
					}
				} else {
					throw new TypeError("State Tracker only accepts strings");
				}
			},
			
			back: function () {
				return stateStack.length > 0 ?
					stateStack.pop() : EMPTY;
			}
		};
	};
	
	ns.StateTracker.EMPTY = EMPTY;
}(DUB.model));

File: Assets/src/StateTracker.cs
using UnityEngine;
using System.Collections.Generic;

public class StateTracker {
	private const int MAX_STATES = 250;
	private Stack<string> stateStack;
	
	public static string EMPTY = "EMPTY";
	
	public StateTracker() {
		this.stateStack = new Stack<string>();
	}
	
	public void Forward(string state) {
		stateStack.Push(state);
		if (stateStack.Count > 250) {
			//stateStack.shift();
		}
	}
			
	public string Back() {
		return stateStack.Count > 0 ? stateStack.Pop() : EMPTY;
	}
}
